pkg/minikube/update/update.go

```diff
@@ -162,40 +180,73 @@ func updateBinary(v semver.Version, downloadBinary, updateLinkPrefix, downloadLi
 		os.Exit(1)
 	}
 
-	url := fmt.Sprintf(downloadLinkFormat, v, downloadBinary)
-	updateBinaryFile(url, checksum)
+	url := fmt.Sprintf(downloadLinkFormat, v, downloadArchive)
+	err = updateBinaryFile(url, checksum)
+	if err != nil {
+		return err
+	}
 
 	env := os.Environ()
-	args := os.Args
+	args := []string{currentBinary, "version"}
 	err = syscall.Exec(currentBinary, args, env)
 	if err != nil {
 		glog.Errorf("Failed to execute updated binary %s: %s", currentBinary, err)
 		os.Exit(1)
 	}
+	return nil
 }
 
-func updateBinaryFile(url string, checksum []byte) {
-	fmt.Println("Downloading updated binary")
+func updateBinaryFile(url string, checksum []byte) error {
+	var binary string
+	fmt.Println("Downloading updated archive")
 	httpResp, err := http.Get(url)
 	if err != nil {
-		glog.Errorf("Cannot download binary: %s", err)
+		glog.Errorf("Cannot download archive: %s", err)
 		os.Exit(1)
 	}
 	defer func() { _ = httpResp.Body.Close() }()
 
-	binary := httpResp.Body
+	updatedArchive := httpResp.Body
 	if httpResp.ContentLength > 0 {
 		bar := pb.New64(httpResp.ContentLength).SetUnits(pb.U_BYTES)
 		bar.Start()
-		binary = bar.NewProxyReader(binary)
+		updatedArchive = bar.NewProxyReader(updatedArchive)
 		defer func() {
 			<-time.After(bar.RefreshRate)
 			fmt.Println()
 		}()
 	}
-	err = update.Apply(binary, update.Options{
-		Hash:     crypto.SHA256,
-		Checksum: checksum,
+
+	// verify checksum
+	archiveBytes, err := ioutil.ReadAll(updatedArchive)
+
+	archiveChecksum, err := checksumFor(crypto.SHA256, archiveBytes)
+
+	if !bytes.Equal(archiveChecksum, checksum) {
+		fmt.Println(fmt.Errorf("Updated file has wrong checksum. Expected: %x, got: %x", archiveChecksum, checksum))
+		os.Exit(1)
+	}
+
+	// Create a temporary directory inside minishift directory to store archive contents
+
+	dir, err := ioutil.TempDir(constants.Minipath, "tmp/download")
+	if err != nil {
+		return err
+		// atexit.ExitWithMessage(1, "Could not create a temporary directory to store archive contents")
+	}
+
+	defer os.RemoveAll(dir) // clean up
+
+	binary = extractAndReplace(runtime.GOOS, dir, archiveBytes)
```

Comment:
> I am not sure why we need to pass `archiveBytes` instead of `updatedArchive`

My understanding is that we need to work on `[] byte` for writing and unpacking the binary from tar/zip. So instead of having to do `archiveBytes, err := ioutil.ReadAll(updatedArchive)` again, I passed it to the function. Is that a bad idea or is their a better way to do it?

> Why _ extractAndReplace_, what does get replaced there?

Agree with you. Renaming it to `extractBinary` since it's returning a binary. Does that sound OK?